src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,908 @@
+    for (size_t ry = 0, ey = image_props.m_canvas_height; ry < ey; ++ry)
+    {
+        for (size_t rx = 0, ex = image_props.m_canvas_width; rx < ex; ++rx)
+        {
+                impl->m_image->set_pixel(rx, ry, pixel_values.data());
+        }
+    }
+    impl->m_tile_name_map.clear();
+}
+
+auto_release_ptr<AOVAccumulator> CryptomatteAOV::create_accumulator() const
+{
+    return auto_release_ptr<AOVAccumulator>(
+        new CryptomatteAOVAccumulator(
+            *impl->m_image,
+            impl->m_pixel_samples,
+            impl->m_tile_name_map,
+            impl->m_num_layers,
+            impl->m_layer_type));
+}
+
+bool CryptomatteAOV::write_images(
+    const char*             file_path,
+    const ImageAttributes&  image_attributes) const
+{
+    const bf::path boost_file_path(file_path);
+    const bf::path directory = boost_file_path.parent_path();
+    const string base_file_name = boost_file_path.stem().string();
+    const string extension = boost_file_path.extension().string();
+
+    const string exr_file_name = base_file_name + ".cryptomatte" + extension;
+    const string exr_file_path = (directory / exr_file_name).string();
+
+    string layer_name;
+
+    switch (impl->m_layer_type)
+    {
+      case CryptomatteType::ObjectNames:
+        layer_name = "AppleseedObjectName";
+        break;
+
+      case CryptomatteType::MaterialNames:
+        layer_name = "AppleseedMaterialName";
+        break;
+
+      assert_otherwise;
+    }
+
+    uint32 type_name_hash = 0;
+    MurmurHash3_x86_32(reinterpret_cast<const unsigned char*>(layer_name.c_str()), static_cast<int>(layer_name.size()), 0, &type_name_hash);
+
+    char type_name_hex[9];
+    sprintf(type_name_hex, "%08x", type_name_hash);
+
+    ImageAttributes image_attributes_copy(image_attributes);
+    image_attributes_copy.insert("color_space", "linear");
+
+    string layer_prefix("cryptomatte/");
+    layer_prefix.append(string(type_name_hex).erase(7));
+    image_attributes_copy.insert(layer_prefix + "/conversion", "uint32_to_float32");
+    image_attributes_copy.insert(layer_prefix + "/hash", "MurmurHash3_32");
+    image_attributes_copy.insert(layer_prefix + "/name", layer_name);
+
+    NameMap name_map;
+    for (const auto& pair : impl->m_tile_name_map)
+    {
+        for (const auto& hash : pair.second)
+        {
+            if (hash.first == 0)
+                continue;
+            name_map[hash.first] = hash.second;
+        }
+    }
+
+    string manifest_str;
+    for (auto& hash : name_map)
+    {
+        char hash_hex[9];
+        sprintf(hash_hex, "%08x", hash.first);
+        manifest_str.append(format("\"{0}\":\"{1}\",", hash.second, string(hash_hex)));
+    }
+    if (manifest_str.size() > 0)
+        manifest_str.pop_back();
+
+    const string manifest_string = format("{{0}}", manifest_str);
+
+    image_attributes_copy.insert(layer_prefix + "/manifest", manifest_string.c_str());
+
+    vector<string> channel_names;
```

Comment:
You can reserve a number of items here since you already know how many you'll push.